etc/docker-scripts/bootstrapper.py

```diff
@@ -118,6 +122,150 @@ def process_outputs(self) -> None:
         duration = time.time() - t0
         OpUtil.log_operation_info('outputs processed', duration)
 
+    def process_metrics(self) -> None:
+        """Expose metrics and    metadata in the KFP UI
+
+        When invoked this method produces a
+        https://www.kubeflow.org/docs/pipelines/sdk/output-viewer
+        compatible metadata file, which esposes the following information
+        for each pipeline node in the KFP UI:
+         - the name of the input file (e.g. notebook.ipynb)
+         - the name of the input archive (e.g. notebook...tar.gz)
+         - the location of the input archive on COS
+
+        If the notebook | script produced a metrics file it is also
+        exposed in the KFP UI:
+        https://www.kubeflow.org/docs/pipelines/sdk/pipelines-metrics/
+
+        There should be no need to override this method in subclasses
+        because the code is not file type specific.
+        """
+
+        OpUtil.log_operation_info('processing metrics and metadata')
+        t0 = time.time()
+
+        # Location where the KFP specific output files will be stored
+        output_path = Path('/tmp')
+
+        # Name of the proprietary KFP UI metadata file.
+        # Each NotebookOp must declare this as an output file or
+        # the KFP UI won't pick up the information.
+        kfp_ui_metadata_filename = 'mlpipeline-ui-metadata.json'
+
+        try:
+            # try to load the metadata file, if one was produced by the
+            # notebok or script
+            src = Path('.') / kfp_ui_metadata_filename
+            logger.debug('Loading UI metadata from {}'.format(src))
+            with open(src, 'r') as f:
+                metadata = json.load(f)
+        except FileNotFoundError:
+            # The script | notebook didn't produce the file
+            # we are looking for.
+            logger.debug('{} produced no file named {}'
+                         .format(self.filepath,
+                                 src))
+            metadata = {}
+        except Exception as ex:
+            # Something is wrong with the user-generated metadata file.
+            logger.warning('Ignoring incompatible {} produced by {}: {} {}'.
+                           format(str(src),
+                                  self.filepath,
+                                  ex,
+                                  str(ex)))
+            metadata = {}
```

Comment:
This code is essentially identical to the block below from lines 219-264 with two exceptions...
1. The first block adds 'UI' to its debug statement
2. The second block adds try/except to validate the JSON.

Could we make this a method, adding an indicator that it's the UI block or just infer that from the filename?  This will reduce a lot of redundant code.

Reply:
I'm not sure that in this case the benefit of consolidating similar code outweighs the drawbacks. My primary concern is that  the consolidated code would be harder to understand because additional conditionals need to be introduced to account for the differences in how the files are processed ("create/append to file") vs "copy file".

Reply:
Sorry, I hadn't noticed the save file operation amongst the error handling that I had focused on as being redundant.  I guess I'd have placed the [save KFP file code](https://github.com/elyra-ai/kfp-notebook/pull/64/files#diff-d2a1666eae3e938008e9735552b3c1a87bf49625300930a6c935fef09ff1dadaR240-R246) after the method when called for the metadata file (for the copy), just like you do for the create/append step on the UI file.

Reply:
I'll try to consolidate

Reply:
done